### PR TITLE
[Bugfix][Spec Decode] Fix GLM-5 MTP low per-position acceptance rate

### PR DESCRIPTION
**Pull request by theanh-ktmt**
## Purpose

GLM-5 (`GlmMoeDsaForCausalLM`, `model_type: glm_moe_dsa`) reuses the DeepSeek MTP module (`DeepSeekMTP`) for speculative decoding. With `num_speculative_tokens > 1`, per-position acceptance collapses after position 0 (e.g. position 2 ≈ 10%).

**The Situation:** 
- GLM NextN applies the shared-head RMSNorm at the **end of every draft step**, so the hidden state carried into the next step is normalized. 
- DeepSeek-V3/V3.2 instead carries the **raw** (pre-norm) hidden and normalizes once in `compute_logits`. 

Because GLM-5 runs through the shared DeepSeek path, the hidden carried into draft steps 1+ was never normalized, leading to mis-scaling the input to the next prediction layer and destroying acceptance for positions 1 and above.

> This is a silent quality bug: no crash or error, only degraded acceptance rate.

## The Fix

`vllm/model_executor/models/deepseek_mtp.py`:

- Detect GLM NextN once via the main model's `model_type == "glm_moe_dsa"`.
- Apply the shared-head norm at the end of `forward` so the carried hidden is normalized.
- Skip the now-redundant norm in `compute_logits` to avoid double-normalizing.

DeepSeek-V3/V3.2 take the original path and are untouched.

## Results

ShareGPT, 256 prompts, `sharegpt-output-len 1024`, `num_speculative_tokens=3`.

**GLM-5-FP8** (fixed by this PR):

| | Acceptance rate | Acceptance length | Pos 0 | Pos 1 | Pos 2 |
|---|---|---|---|---|---|
| Before | 39.95% | 2.20 | 71.38% | 38.08% | 10.39% |
| After  | **49.36%** | **2.48** | 70.32% | **46.93%** | **30.83%** |

**DeepSeek-V3.2** (expected unchanged):

| | Acceptance rate | Acceptance length | Pos 0 | Pos 1 | Pos 2 |
|---|---|---|---|---|---|
| Before | 41.29% | 2.24 | 68.52% | 37.90% | 17.44% |
| After  | 41.32% | 2.24 | 68.73% | 37.87% | 17.35% |

**Comments:**
- Position 0 is unaffected (it consumes the target model's hidden).
- The fix recovers positions 1/2, which depend on the MTP module's carried hidden.
- DeepSeek V3.2 acceptance is unaffected.

## Environment

- Image: `vllm/vllm-openai-rocm:v0.23.0`
- Machine: AMD MI325X (gfx942) ×8
- OS: Ubuntu 22.04.5
- Python Version: 3.12
- Models: `zai-org/GLM-5-FP8` (FP8), `deepseek-ai/DeepSeek-V3.2`
- Dataset: [ShareGPT_V3_unfiltered_cleaned_split.json](https://huggingface.co/datasets/anon8231489123/ShareGPT_Vicuna_unfiltered/blob/main/ShareGPT_V3_unfiltered_cleaned_split.json)

## Reproduce

Serve models:

```bash
# GLM-5
VLLM_ROCM_USE_AITER=1 vllm serve zai-org/GLM-5-FP8 \
    --trust-remote-code --tensor-parallel-size 8 \
    --reasoning-parser glm45 --tool-call-parser glm47 --enable-auto-tool-choice \
    --speculative-config '{"method": "mtp", "num_speculative_tokens": 3}'

# DeepSeek-V3.2
VLLM_ROCM_USE_AITER=1 vllm serve deepseek-ai/DeepSeek-V3.2 \
    --trust-remote-code --tensor-parallel-size 8 \
    --speculative-config '{"method": "mtp", "num_speculative_tokens": 3}'
```

Benchmark:

```bash
MODEL=zai-org/GLM-5-FP8   # or deepseek-ai/DeepSeek-V3.2
vllm bench serve \
    --model "$MODEL" --backend vllm --base-url http://localhost:8000 \
    --dataset-name sharegpt \
    --dataset-path ShareGPT_V3_unfiltered_cleaned_split.json \
    --num-prompts 256 --sharegpt-output-len 1024
```

## Test

- Lint check: **Pass**
- End-to-end serve + `vllm bench serve` for both models: **Table above**
- Expected behaviors:
    - GLM-5 acceptance **improved**
    - DeepSeek-V3.2 acceptance **unchanged**

### CODE REVIEW

**Comment by theanh-ktmt** [comment]
Friendly ping for review 🙏 

This fixes GLM-5 MTP low per-position acceptance by applying the GLM per-step shared-head norm in the shared DeepSeek MTP path.

Results (ShareGPT, 256 prompts, num_speculative_tokens=3):
- GLM-5-FP8: improved acceptance length 2.20 → 2.48, pos-2 acceptance 10.4% → 30.8%
- DeepSeek-V3.2: unchanged (2.24 → 2.24)

@luccafong, could you take a look, and add the `ready` label to kick off CI? 

First-time contributor, so tests are gated. Thanks!

**Comment by theanh-ktmt** [comment]
Closing this — the same root cause is already fixed on `main` by #45895 (commit `ab6660699`), which recycles the post-final-norm hidden state into the next MTP draft step (matching SGLang's `deepseek_nextn`). That fix is applied to all DeepSeek MTP models and resolves the GLM-5 low per-position acceptance reported here, so this PR is redundant. Thanks!
